resources/certificate.rb

```diff
@@ -65,36 +53,83 @@
 end
 
 action :delete do
-  # do we have a hash or a subject?
-  # TODO: It's a bit annoying to know the thumbprint of a cert you want to remove when you already
-  # have the file.  Support reading the hash directly from the file if provided.
-  search = if new_resource.source =~ /^[a-fA-F0-9]{40}$/
-             "Thumbprint -eq '#{new_resource.source}'"
-           else
-             "Subject -like '*#{new_resource.source.sub(/\*/, '`*')}*'" # escape any * in the source
-           end
-  cert_command = "Get-ChildItem Cert:\\#{cert_location}\\#{new_resource.store_name} | where { $_.#{search} }"
-
-  code_script = within_store_script do |store|
-    <<-EOH
-foreach ($c in #{cert_command})
-{
-  #{store}.Remove($c)
-}
-EOH
+  delete_cert_from_certstore
+end
+
+action :fetch do
+  cert_obj = fetch_cert_from_certstore
+  if cert_obj
+    show_or_store_cert(cert_obj)
+  else
+    Chef::Log.info('Certificate not found')
   end
-  guard_script = "@(#{cert_command}).Count -gt 0\n"
-  powershell_script "Removing certificate #{new_resource.source} from #{cert_location}\\#{new_resource.store_name}" do
-    guard_interpreter :powershell_script
-    convert_boolean_return true
-    code code_script
-    only_if guard_script
+end
+
+action :verify do
+  out = verify_cert_from_certstore
+  if !!out == out
+    out = out ? 'Certificate is valid' : 'Certificate not valid'
   end
+  Chef::Log.info(out.to_s)
 end
 
 action_class do
   include Windows::Helper
 
+  def add_cert_in_certstore
+    add_cert(openssl_cert_obj)
+  end
+
+  def delete_cert_from_certstore
+    delete_cert
+  end
+
+  def fetch_cert_from_certstore
+    fetch_cert
+  end
+
+  def verify_cert_from_certstore
+    verify_cert
+  end
+
+  def show_or_store_cert(cert_obj)
+    if cert_path
+      export_cert(cert_obj, cert_path)
+      if ::File.size(cert_path) > 0
+        Chef::Log.info("Certificate export in #{cert_path}")
+      else
+        ::File.delete(cert_path)
+      end
+    else
+      Chef::Log.info(cert_obj.display)
+    end
+  end
+
+  def export_cert(cert_obj, cert_path)
+    out_file = ::File.new(cert_path, 'w+')
+    case ::File.extname(cert_path)
+    when '.pem'
+      out_file.puts(cert_obj.to_pem)
+    when '.der'
+      out_file.puts(cert_obj.to_der)
+    when '.cer'
+      cert_out = powershell_out("openssl x509 -text -inform DER -in #{cert_obj.to_pem} -outform CER").stdout
```

Comment:
if this something we could do in ruby vs. shelling out?

Reply:
yes @tas50 we can achieve that in ruby as well but this is super easy way with less amount of code also if `openssl` command failed due to any reason powershell_out track error easily. Please let me know your thought about it.